### PR TITLE
Fix: Groups support for Secret Approvals and Access Requests

### PR DESCRIPTION
**Pull request by DanielHougaard**

### CODE REVIEW

File: backend/src/db/migrations/20240429175301_fix-db-reference-for-groups-and-project-memberships.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    await knex.schema.alterTable(TableName.SecretApprovalPolicyApprover, (t) => {
+      t.uuid("approverId").nullable().alter();
+
+      // add new "approverUserId" column
+      t.uuid("approverUserId").nullable();
+      t.foreign("approverUserId").references("id").inTable(TableName.Users).onDelete("CASCADE");
+    });
+
+    await knex.schema.alterTable(TableName.SecretApprovalRequest, (t) => {
```
Comment (akhilmhdh) [inline backend/src/db/migrations/20240429175301_fix-db-reference-for-groups-and-project-memberships.ts:25]: I was testing identity for secret replication and issue is then i would need to add all these fields again . May be instead we do it like permission there is non nullable   - `type"which says whether its user or identity and keep rest as generic like before

Then do a conditional join

File: backend/src/ee/services/access-approval-request/access-approval-request-dal.ts [hunk 2]
```diff
@@ -11,6 +11,32 @@ export type TAccessApprovalRequestDALFactory = ReturnType<typeof accessApprovalR
 
 export const accessApprovalRequestDALFactory = (db: TDbClient) => {
   const accessApprovalRequestOrm = ormify(db, TableName.AccessApprovalRequest);
+  const projectUserAdditionalPrivilegeOrm = ormify(db, TableName.ProjectUserAdditionalPrivilege);
+
+  const deleteMany = async (filter: TFindFilter<TAccessApprovalRequests>, tx?: Knex) => {
+    try {
+      const handleDeletion = async (processedTx: Knex) => {
```
Comment (akhilmhdh) [inline backend/src/ee/services/access-approval-request/access-approval-request-dal.ts:18]: You could do a  `const processedTx = tx || (await db.transaction())` and used it like normal without function inside function pattern.

https://knexjs.org/guide/transactions.html

File: backend/src/ee/services/access-approval-request/access-approval-request-dal.ts [hunk 3]
```diff
@@ -11,6 +11,32 @@ export type TAccessApprovalRequestDALFactory = ReturnType<typeof accessApprovalR
...
+        const accessApprovalRequests = await accessApprovalRequestOrm.find(filter, { tx: processedTx });
+
+        await projectUserAdditionalPrivilegeOrm.delete(
+          {
+            $in: {
+              id: accessApprovalRequests.filter((req) => !!req.privilegeId).map((req) => req.privilegeId!)
```
Comment (akhilmhdh) [inline backend/src/ee/services/access-approval-request/access-approval-request-dal.ts:24]: `!!req.privilegeId` => `Boolean(req.privilegeId)`

File: backend/src/db/migrations/20240429175301_fix-db-reference-for-groups-and-project-memberships.ts (hunk 1 above)
Comment (akhilmhdh) [reply backend/src/db/migrations/20240429175301_fix-db-reference-for-groups-and-project-memberships.ts:25]: Kindly make the migration file idempotent

File: backend/src/db/migrations/20240429172301_access_approval_requests.ts [hunk 4]
```diff
@@ -31,21 +37,39 @@ export async function up(knex: Knex): Promise<void> {
...
+      t.foreign("memberUserId").references("id").inTable(TableName.Users).onDelete("CASCADE");
+
       t.string("status").notNullable();
       t.uuid("requestId").notNullable();
       t.foreign("requestId").references("id").inTable(TableName.AccessApprovalRequest).onDelete("CASCADE");
       t.timestamps(true, true, true);
     });
   }
+
+  if (!(await knex.schema.hasColumn(TableName.ProjectUserAdditionalPrivilege, "groupMembershipId"))) {
```
Comment (akhilmhdh) [inline backend/src/db/migrations/20240429172301_access_approval_requests.ts:51]: Kindly create a separate table for `GroupAdditionalPrivilege` as this aligns like `IdentityUserAdditionalPrivilege`

File: backend/src/ee/services/group/group-dal.ts [hunk 5]
```diff
@@ -5,10 +5,75 @@ import { TableName, TGroups } from "@app/db/schemas";
...
+  const groupMembershipOrm = ormify(db, TableName.GroupProjectMembership);
+  const accessApprovalRequestOrm = ormify(db, TableName.AccessApprovalRequest);
+  const secretApprovalRequestOrm = ormify(db, TableName.SecretApprovalRequest);
+
+  const deleteMany = async (filterQuery: TFindFilter<TGroups>, tx?: Knex) => {
+    const handleDeletion = async (processedTx: Knex) => {
+      // Find all memberships
+      const groups = await groupOrm.find(filterQuery, { tx: processedTx });
+
+      for await (const group of groups) {
```
Comment (akhilmhdh) [inline backend/src/ee/services/group/group-dal.ts:23]: I think i need some explanation comments to this one. Didn't understand what's happeninng?

**Review by akhilmhdh — CHANGES_REQUESTED** [review]

File: backend/src/ee/services/group/group-dal.ts (hunk 5 above)
Comment (DanielHougaard) [reply backend/src/ee/services/group/group-dal.ts]: I added comments to the file, I hope that helps!

**Comment by maidul98** [comment]
Already addressed by a previous PR

**Pull request closed without merging**